src/ActionList/List.tsx

```diff
@@ -42,32 +47,42 @@ export const List = React.forwardRef<HTMLUListElement, ActionListProps>(
       paddingY: variant === 'inset' ? 2 : 0,
     }
 
+    const [slots, childrenWithoutSlots] = useSlots(props.children, {
+      heading: Heading,
+    })
```

Comment:
Would this already work with eslint-plugin-primer-react or do we need to update that as well?

Reply:
I think we would need to make some update on the eslint plugin. 

- We need to add `ActionList: ['ActionList.Header'],` in [this object](https://github.com/primer/eslint-plugin-primer-react/blob/main/src/rules/direct-slot-children.js#L5)

Though, this rule will only prevent folks from using `ActionList.Heading` outside of `ActionList`, this won't warn if the `ActionList` is missing `ActionList.Heading` which I believe this is something we need as well? Can we do this with the type system at all? 🤔 Or should we look into adding another eslint rule for that too?

Also we can add `ActionList.Heading` to the [a11y-explicit-heading rule](https://github.com/primer/eslint-plugin-primer-react/blob/main/src/rules/a11y-explicit-heading.js#L7) to restrict the values passed to `as`

Reply:
Considering action lists in menus shouldn't have a heading ([reference](https://github.com/primer/react/pull/3581#discussion_r1312461027)), maybe we can enforece this with the eslint rule as well?

Reply:
> Also we can add ActionList.Heading to the [a11y-explicit-heading rule](https://github.com/primer/eslint-plugin-primer-react/blob/main/src/rules/a11y-explicit-heading.js#L7) to restrict the values passed to as

Because ActionList.Heading is a new component, we can also use typescript to restrict allowed values for `as` :)

Reply:
> Considering action lists in menus shouldn't have a heading (https://github.com/primer/react/pull/3581#discussion_r1312461027), maybe we can enforece this with the eslint rule as well?

Good point. We could add an eslint rule for this, however, I'm not sure how effectively we would be able to catch incorrect usage given ActionList does not have to be a direct child of ActionMenu, it can be abstracted in a custom `UserList` or `BranchList` component that uses ActionList internally.

We also have the option of a dev environment runtime warning. Because [we know when an ActionList is contained inside ActionMenu](https://github.com/primer/react/blob/main/src/ActionList/Item.tsx#L64), we could use that to throw a warning

Reply:
>Because ActionList.Heading is a new component, we can also use typescript to restrict allowed values for as :)

That is so true! I missed that 😅  which is [already restricted](https://github.com/primer/react/blob/5a316b5f8bd61915e8c23f22edd71e749072741d/src/ActionList/Heading.tsx#L11) anyway! 

>We also have the option of a dev environment runtime warning. Because [we know when an ActionList is contained inside ActionMenu](https://github.com/primer/react/blob/main/src/ActionList/Item.tsx#L64), we could use that to throw a warning

I love this! Thanks for the context.  I made it `invariant` to throw an error. Because first, it is a new component and it won't break anything. Second, when there is a heading the list will be labelled by the heading and completely disregard the button name. This could potentially be misleading. 

All that said, if you think that is unnecessary and very little chance of creating a misleading experience, happy to make it warning as well 😊